README.md
geoparser
=========

[![Project Status: Inactive – The project has reached a stable, usable
state but is no longer being actively developed; support/maintenance
will be provided as time
allows.](http://www.repostatus.org/badges/latest/inactive.svg)](http://www.repostatus.org/#inactive)
[![Build
Status](https://travis-ci.org/ropensci/geoparser.svg?branch=master)](https://travis-ci.org/ropensci/geoparser)
[![Build
status](https://ci.appveyor.com/api/projects/status/7sw9ufcgh8pk1r5d?svg=true)](https://ci.appveyor.com/project/ropensci/geoparser)
[![codecov](https://codecov.io/gh/ropensci/geoparser/branch/master/graph/badge.svg)](https://codecov.io/gh/ropensci/geoparser)
[![](https://badges.ropensci.org/43_status.svg)](https://github.com/ropensci/onboarding/issues/43)

This package is an interface to the [geoparser.io
API](https://geoparser.io) that identifies places mentioned in text,
disambiguates those places, and returns data about the places found in
the text.

Installation
============

To install the package, you will need the devtools package.

``` r
library("devtools")
install_github("ropensci/geoparser")
```

To get an API key, you need to register at
<a href="https://geoparser.io/pricing.html" class="uri">https://geoparser.io/pricing.html</a>.
With an hobbyist account, you can make up to 1,000 calls a month to the
API. For ease of use, save your API key as an environment variable as
described at
<a href="https://stat545-ubc.github.io/bit003_api-key-env-var.html" class="uri">https://stat545-ubc.github.io/bit003_api-key-env-var.html</a>.

The package will conveniently look for your API key using
`Sys.getenv("GEOPARSER_KEY")` so if your API key is an environment
variable called “GEOPARSER\_KEY” you don’t need to input it manually.

What is geoparsing?
===================

According to [Wikipedia](https://en.wikipedia.org/wiki/Geoparsing),
geoparsing is the process of converting free-text descriptions of places
(such as “Springfield”) into unambiguous geographic identifiers (such as
lat-lon coordinates). A geoparser is a tool that helps in this process.
Geoparsing goes beyond geocoding in that, rather than analyzing
structured location references like mailing addresses and numerical
coordinates, geoparsing handles ambiguous place names in unstructured
text.

Geoparser.io works best on complete sentences in *English*. If you have
a very short text, such as a partial address like
“`Auckland New Zealand`,” you probably want to use a geocoder tool
instead of a geoparser. In R, you can use the
[opencage](https://cran.r-project.org/package=opencage) package for
geocoding!

How to use the package
======================

You need to input a text whose size is less than 8KB.

``` r
library("geoparser")
output <- geoparser_q("I was born in Vannes and I live in Barcelona")
```

The output is list of 2 `data.frame`s (`dply::tbl_df`s). The first one
is called `properties` and contains

-   the api version called `apiVersion`

-   the `source` of the results

-   the `id` of the query

-   `text_md5` is the MD5 hash of the text that was sent to the API.

``` r
output$properties
```

    ## # A tibble: 1 x 4
    ##   apiVersion source       id                  text_md5                     
    ##   <fct>      <fct>        <fct>               <chr>                        
    ## 1 0.5.2      geoparser.io eVYeJaJuoMZRuAap5e… 51e05aeb3366e55795a9729dd74a…

The second data.frame contains the results and is called results:

``` r
knitr::kable(output$results)
```

| country | confidence | name      | admin1 | type                                           | geometry.type |  longitude|  latitude|  reference1|  reference2| text\_md5                        |
|:--------|:-----------|:----------|:-------|:-----------------------------------------------|:--------------|----------:|---------:|-----------:|-----------:|:---------------------------------|
| FR      | 1          | Vannes    | A2     | seat of a second-order administrative division | Point         |   -2.75000|  47.66667|          14|          20| 51e05aeb3366e55795a9729dd74ae901 |
| ES      | 1          | Barcelona | 56     | seat of a first-order administrative division  | Point         |    2.15899|  41.38879|          35|          44| 51e05aeb3366e55795a9729dd74ae901 |

-   `country` is the [ISO-3166 2-letter country
    code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) for the
    country in which this place is located, or NULL for features outside
    any sovereign territory.

-   `confidence` is a confidence score produced by the place name
    disambiguation algorithm. Currently returns a placeholder value;
    subject to change.

-   `name` is the best name for the specified location, with a
    preference for official/short name forms (e.g., “`New York`” over
    “`NYC`,” and “`California`” over “`State of California`”), which may
    be different from exactly what appears in the text.

-   `admin1` is a code representing the state/province-level
    administrative division containing this place. (From GeoNames.org:
    *“Most adm1 are FIPS codes. ISO codes are used for US, CH, BE and
    ME. UK and Greece are using an additional level between country and
    fips code. The code ‘`00`’ stands for general features where no
    specific adm1 code is defined.”*).

-   `type` is a text description of the geographic feature type — see
    &lt;GeoNames.org&gt; for a complete list. Subject to change.

-   `geometry.type` is the type of the geographical feature, e.g.
    “`Point`”.

-   `longitude` is the longitude.

-   `latitude` is the latitude.

-   `reference1` is the start (index of the first character in the place
    reference) – each reference to this place name found in the input
    text is on one distinct line.

-   `reference2` the end (index of the first character after the place
    reference) – each reference to the place name found in the input
    text is on one distinct line.

-   `text_md5` is the MD5 hash of the text that was sent to the API.

You can input a vector of characters since the function is vectorized.
This is the case where the MD5 hash of each text can be useful for
further analysis.

``` r
library("geoparser")
output_v <- geoparser_q(text_input = c("I was born in Vannes but I live in Barcelona.",
"France is the most beautiful place in the world.", "No place here."))
knitr::kable(output_v$results)
```

| country | confidence | name      | admin1 | type                                           | geometry.type |  longitude|  latitude|  reference1|  reference2| text\_md5                        |
|:--------|:-----------|:----------|:-------|:-----------------------------------------------|:--------------|----------:|---------:|-----------:|-----------:|:---------------------------------|
| FR      | 1          | Vannes    | A2     | seat of a second-order administrative division | Point         |   -2.75000|  47.66667|          14|          20| 90aba603d6b3f6b916c634f74ebc3a05 |
| ES      | 1          | Barcelona | 56     | seat of a first-order administrative division  | Point         |    2.15899|  41.38879|          35|          44| 90aba603d6b3f6b916c634f74ebc3a05 |
| FR      | 1          | France    | 00     | independent political entity                   | Point         |    2.00000|  46.00000|           0|           6| 33247ffc493ca57619549e512c7b5c59 |

``` r
knitr::kable(output_v$properties)
```

| apiVersion | source       | id                    | text\_md5                        |
|:-----------|:-------------|:----------------------|:---------------------------------|
| 0.5.2      | geoparser.io | BDx1bAbcrXV3u5WXaB62y | 90aba603d6b3f6b916c634f74ebc3a05 |
| 0.5.2      | geoparser.io | eVYeJaJuoMZRuAap5eQra | 33247ffc493ca57619549e512c7b5c59 |
| 0.5.2      | geoparser.io | WAWBZdZhwOEQU4YaJNLb5 | a9b35a32dc022502c943daa55520bfc0 |

How does it work?
=================

The API uses the Geonames.org gazetteer data. Geoparser.io uses a
variety of named entity recognition tools to extract location names from
the raw text input, and then applies a proprietary disambiguation
algorithm to resolve location names to specific gazetteer records.

What happens if the same place occurs several times in the text?
================================================================

If the input text contains several times the same placename, there is
one line for each repetition, the only difference between lines being
the values of `reference1` and `reference2`.

``` r
output2 <- geoparser_q("I like Paris and Paris and Paris and yeah it is in France!")
knitr::kable(output2$results)
```

| country | confidence | name   | admin1 | type                          | geometry.type |  longitude|  latitude|  reference1|  reference2| text\_md5                        |
|:--------|:-----------|:-------|:-------|:------------------------------|:--------------|----------:|---------:|-----------:|-----------:|:---------------------------------|
| FR      | 1          | France | 00     | independent political entity  | Point         |     2.0000|  46.00000|          51|          57| 34ac61cd71faef0cc4b336b706a7e545 |
| FR      | 1          | Paris  | A8     | capital of a political entity | Point         |     2.3488|  48.85341|           7|          12| 34ac61cd71faef0cc4b336b706a7e545 |
| FR      | 1          | Paris  | A8     | capital of a political entity | Point         |     2.3488|  48.85341|          17|          22| 34ac61cd71faef0cc4b336b706a7e545 |
| FR      | 1          | Paris  | A8     | capital of a political entity | Point         |     2.3488|  48.85341|          27|          32| 34ac61cd71faef0cc4b336b706a7e545 |

What happens if there are no results for the text?
==================================================

In this case the results table is empty.

``` r
output_nothing <- geoparser_q("No placename can be found.")
output_nothing$results
```

    ## # A tibble: 0 x 1
    ## # … with 1 variable: text_md5 <chr>

How well does it work?
======================

The API team has tested the API un-scientifically and noticed a
performance similar to other existing geoparsing tools. A scientific
evaluation is under way. The public Geoparser.io API works best with
professionally-written, professionally-edited news articles, but for
Enterprise customers the API team says that it can be tuned/tweaked for
other kinds of input (e.g., social media).

Let’s look at this example:

``` r
output3 <- geoparser_q("I live in Hyderabad, India. My mother would prefer living in Hyderabad near Islamabad!")
knitr::kable(output3$results)
```

| country | confidence | name       | admin1 | type                                          | geometry.type |  longitude|  latitude|  reference1|  reference2| text\_md5                        |
|:--------|:-----------|:-----------|:-------|:----------------------------------------------|:--------------|----------:|---------:|-----------:|-----------:|:---------------------------------|
| IN      | 1          | Hyderabad  | 40     | seat of a first-order administrative division | Point         |   78.45636|  17.38405|          10|          19| 645d890dde2bce1092338f0cbc7af011 |
| IN      | 1          | Hyderabad  | 40     | seat of a first-order administrative division | Point         |   78.45636|  17.38405|          61|          70| 645d890dde2bce1092338f0cbc7af011 |
| IN      | 1          | India      | 00     | independent political entity                  | Point         |   79.00000|  22.00000|          21|          26| 645d890dde2bce1092338f0cbc7af011 |
| BD      | 1          | Chittagong | 84     | seat of a first-order administrative division | Point         |   91.83168|  22.33840|          76|          85| 645d890dde2bce1092338f0cbc7af011 |

Geoparser.io typically assumes two mentions of the same name appearing
so closely together in the same input text refer to the same place. So,
because it saw “`Hyderabad`” (India) in the first sentence, it assumes
“`Hyderabad`” in the second sentence refers to the same city. Also,
“`Islamabad`” is an alternate name for Chittagong, which has a higher
population than Islamabad (Pakistan) and is closer to Hyderabad (India).

Here is another example with a longer text.

``` r
text <- "Aliwagwag is situated in the Eastern Mindanao Biodiversity \
Corridor which contains one of the largest remaining blocks of tropical lowland \
rainforest in the Philippines. It covers an area of 10,491.33 hectares (25,924.6 \
acres) and a buffer zone of 420.6 hectares (1,039 acres) in the hydrologically \
rich mountainous interior of the municipalities of Cateel and Boston in Davao \
Oriental as well as a portion of the municipality of Compostela in Compostela \
Valley. It is also home to the tallest trees in the Philippines, the Philippine \
rosewood, known locally as toog. In the waters of the upper Cateel River, a rare \
species of fish can be found called sawugnun by locals which is harvested as a \
delicacy." 

output4 <- geoparser_q(text)
knitr::kable(output4$results)
```

| country | confidence | name                       | admin1 | type                                 | geometry.type |  longitude|  latitude|  reference1|  reference2| text\_md5                        |
|:--------|:-----------|:---------------------------|:-------|:-------------------------------------|:--------------|----------:|---------:|-----------:|-----------:|:---------------------------------|
| PH      | 1          | Philippines                | 0      | independent political entity         | Point         |   122.0000|  13.00000|         159|         170| d89e347a998b58c6a8e54bc9f9abc073 |
| PH      | 1          | Philippines                | 0      | independent political entity         | Point         |   122.0000|  13.00000|         513|         524| d89e347a998b58c6a8e54bc9f9abc073 |
| PH      | 1          | Cateel                     | 11     | populated place                      | Point         |   126.4533|   7.79139|         354|         360| d89e347a998b58c6a8e54bc9f9abc073 |
| PH      | 1          | Boston                     | 11     | populated place                      | Point         |   126.3642|   7.87111|         365|         371| d89e347a998b58c6a8e54bc9f9abc073 |
| PH      | 1          | Province of Davao Oriental | 11     | second-order administrative division | Point         |   126.3333|   7.16667|         375|         390| d89e347a998b58c6a8e54bc9f9abc073 |
| PH      | 1          | Compostela                 | 11     | second-order administrative division | Point         |   126.1167|   7.68333|         435|         445| d89e347a998b58c6a8e54bc9f9abc073 |
| PH      | 1          | Compostela Valley          |        | valley                               | Point         |   125.9586|   7.60755|         449|         467| d89e347a998b58c6a8e54bc9f9abc073 |
| PH      | 1          | Cateel River               | 11     | stream                               | Point         |   126.4533|   7.78750|         602|         614| d89e347a998b58c6a8e54bc9f9abc073 |

What can I do with the results?
===============================

You might want to map them using
[leaflet](https://rstudio.github.io/leaflet/) or
[ggmap](https://cran.r-project.org/package=ggmap) or anything you like.
The API website provides [suggestions of
use](https://geoparser.io/uses.html) for inspiration.

Meta
----

-   Please [report any issues or
    bugs](https://github.com/ropensci/geoparser/issues).
-   License: GPL
-   Get citation information for `geoparser` in R doing
    `citation(package = 'geoparser')`
-   Please note that this project is released with a [Contributor Code
    of
    Conduct](https://github.com/ropensci/geoparser/blob/master/CONDUCT.md).
    By participating in this project you agree to abide by its terms.

[![ropensci\_footer](http://ropensci.org/public_images/github_footer.png)](http://ropensci.org)


NEWS.md
# geoparser (development version)

# geoparser 0.1.2

* Fix error caused by a namespace issue

* Change the behavior of geoparser_key() such that if no key is provided and
no key is saved in .Renviron either, the function errors with an informative 
error message (#11).

# geoparser 0.1.1

* Updates links to the Github repo (moved from ropenscilabs to ropensci)


tests/testthat.R
library(testthat)
library(geoparser)

if (identical(tolower(Sys.getenv("NOT_CRAN")), "true")) {
  test_check("geoparser")
}


tests/testthat/test_query.R
context("query")
test_that("query returns a list of data.frames",{
  testthat::skip_on_cran()
  output <- geoparser_q(text_input = "Paris o Paris and Paris")
  expect_is(output, "list")
  expect_is(output$properties, "tbl_df")
  expect_is(output$results, "tbl_df")
  expect_is(output$results$longitude, "numeric")
  expect_is(output$results$latitude, "numeric")
  expect_is(output$results$reference1, "numeric")
  expect_is(output$results$reference2, "numeric")
  expect_true(nrow(output$results) > 1)
  output <- geoparser_q(text_input = c("Paris o Paris and Paris", "Soon I will travel to Munich" , "I live in Barcelona"))
  expect_is(output, "list")
  expect_is(output$properties, "tbl_df")
  expect_is(output$results, "tbl_df")
  expect_is(output$results$longitude, "numeric")
  expect_is(output$results$latitude, "numeric")
  expect_is(output$results$reference1, "numeric")
  expect_is(output$results$reference2, "numeric")
  expect_true(nrow(output$results) > 1)
})

test_that("no problems if no results",{
  testthat::skip_on_cran()
  output <- geoparser_q(text_input = "no placename here")
  expect_is(output, "list")
  expect_is(output$properties, "tbl_df")
  expect_is(output$results, "tbl_df")
})



tests/testthat/test_checks.R
context("checks of parameters")

test_that("the key has to be a character",{
  testthat::skip_on_cran()
  expect_error(geoparser_q(key = 1),
               "Key should be a character")
})

test_that("the text cannot be bigger than 8kB",{
  testthat::skip_on_cran()
  text <- toString(rep("this should get very long", 500))
  expect_error(geoparser_q(text_input = text),
               "The size of text_input should be smaller than")
  expect_error(geoparser_q(text_input = c("lala", text)),
               "The size of text_input should be smaller than")
})


R/utils.R
#' @importFrom httr content POST add_headers accept_json status_code
#' @importFrom jsonlite fromJSON
#' @importFrom tidyr unite_
#' @importFrom dplyr "%>%" group_by mutate_ select_ ungroup tbl_df rename_ arrange_
#' @importFrom lazyeval interp
#' @importFrom utils URLencode
#' @importFrom purrr map map_df map_dbl safely
#' @importFrom stringr str_split
#' @importFrom digest digest

# status check
#' @noRd
geoparser_check <- function(req) {
  status <- httr::status_code(req)
  if (status < 400) return(invisible())
  stop("HTTP failure: ", status, call. = FALSE)
}


# function that checks the query
#' @noRd
geoparser_query_check <- function(text_input, key){
  # check key
  if(!is.null(key)){
    if(!is.character(key)){
      stop(call. = FALSE, "Key should be a character.")
    }
  }

  # check text
  if(any(nchar(text_input) >= 8192)){
    stop("The size of text_input should be smaller than 8KB.", call. = FALSE)
    }

}

# parse results
#' @noRd
geoparser_parse <- function(req) {
  text <- httr::content(req, as = "text",
                        encoding = "UTF-8")
  if (identical(text, "")) stop("No output to parse",
                                call. = FALSE)
  temp <- jsonlite::fromJSON(text,
                             simplifyVector = FALSE)
  # if we have something to process
  # getting from the raw output to a nice data.frame
  if(length(temp$features) != 0){

    results <- purrr::map(temp$features, unlist)
    results <- suppressWarnings(purrr::map(results, as.data.frame))
    results <- purrr::map(results, t)
    results <- suppressWarnings(purrr::map_df(results, as.data.frame))
    # making coordinates numeric
    results <- results %>%
      mutate_(geometry.coordinates2 = lazyeval::interp(
        ~ as.numeric(as.character(geometry.coordinates2))))
    results <- results %>%
      mutate_(geometry.coordinates1 = lazyeval::interp(
        ~ as.numeric(as.character(geometry.coordinates1))))

    # start modification of references to possibly multiple
    # occurrences of words
    # for this I first put all start and end references together
    # separated by "_"
    which_ref <- which(grepl("references", names(results)))
    first_ind <- which_ref[which(which_ref %% 2 == 1)]
    results <- tidyr::unite_(results,
                             "start",
                             names(results)[first_ind])
    which_ref <- which(grepl("references", names(results)))
    results <- tidyr::unite_(results,
                             "end",
                             names(results)[which_ref])
    # end of the transformation for having 1 line per occurence
    results <- suppressMessages(function_df(results))

    # make names nicer by erasing the "properties." they have
    # at the beginning
    names(results) <- gsub("properties\\.", "", names(results))
    results <- results[, 3:ncol(results)]
  }else{
    results <- dplyr::tbl_df(data.frame(NULL))
  }


  list(properties = dplyr::tbl_df(as.data.frame(temp$properties)),
       results = results)
}

#' @noRd
function_na <- function(vec){
  # sep by _
  temp <- stringr::str_split(vec, "_")[[1]]
  # to numeric, warnings when NA
  temp <- suppressWarnings(as.numeric(temp))
  # count number of not NA
  sum(!is.na(temp))
}

# for CRAN
start <- NULL

# function for transforming start and end
# This was needed because when a word is found several times,
# I want a line per occurence instead of one line with
# the starts and ends of all occurences in one cell.
# I think it's better for further processing.
#' @noRd
function_df <- function(df){
  # arrange else lengths do not correspond to df
  df <- arrange_(df, ~ start)
  lengths <- dplyr::select_(df, "start")
  lengths <- split(lengths, lengths$start)
  lengths <-  purrr::map_dbl(lengths, function_na)

  df <- df[rep(1:nrow(df), lengths), ]
  df <- dplyr::group_by(df, start)
  df <- dplyr::mutate_(df, number = lazyeval::interp(quote(seq_len(dplyr::n()))))
  df <- dplyr::group_by(df, start)
  df <- dplyr::mutate_(df, reference1 = lazyeval::interp(
      quote(
        as.numeric(strsplit(start[1], "_")[[1]][number]))))
  df <- dplyr::mutate_(df, reference2 = interp(
      df <- quote(
        as.numeric(strsplit(end[1], "_")[[1]][number]))))
  df <- dplyr::select_(df, lazyeval::interp(quote(- start)))
  df <- dplyr::select_(df, lazyeval::interp(quote(- id)))
  df <- dplyr::select_(df, lazyeval::interp(quote(- number)))
  df <- dplyr::select_(df, lazyeval::interp(quote(- end)))
  df <- dplyr::ungroup(df)

  df %>%
    dplyr::rename_(longitude =
                     lazyeval::interp(quote(geometry.coordinates1))) %>%
    dplyr::rename_(latitude = lazyeval::interp(quote(geometry.coordinates2)))

}

# base URL for all queries
#' @noRd
geoparser_url <- function() {
  "https://geoparser.io/api/geoparser/"
}

# get results
#' @noRd
geoparser_get <- function(query_par){
  httr::POST(geoparser_url(),
             httr::accept_json(),
             httr::add_headers(
               "Authorization" = paste("apiKey", query_par$apiKey),
               "Content-Type" =
                 "application/x-www-form-urlencoded; charset=UTF-8"
             ),
             body = paste0("inputText=",
                           utils::URLencode(query_par$inputText))
  )
}
# get results
#' @noRd
geoparser_get_safe <- purrr::safely(geoparser_get)

#' Retrieve Geoparser.io API key
#'
#' A Geoparser.io API Key
#' Looks in env var \code{GEOPARSER_KEY}. The function errors if no key is
#' saved in .Renviron.
#'
#' @keywords internal
#' @export
geoparser_key <- function(quiet = TRUE) {
  pat <- Sys.getenv("GEOPARSER_KEY")
  if (identical(pat, ""))  {
    stop("Please provide your Geoparser.io API Key as an argument,
      or save it in .Renviron,
      and if you don't have a key yet register at Geoparser.io.")
  }
  if (!quiet) {
    message("Using Geoparser.io API Key from envvar GEOPARSER_KEY")
  }
  return(pat)
}

#############
# vectorizing
#' @noRd
total <- function(text, key){
  # res
  temp <- geoparser_get_safe(query_par = list(inputText = URLencode(text),
                                         apiKey = key))
  if(!is.null(temp$error)){
    message(paste0("The API call failed, the error message is ", temp$error))
    return(NULL)
  }else{
    temp <- temp$result
  }
  # check message
  geoparser_check(temp)

  # parse
  parsed <- geoparser_parse(temp)
  # add text for future reference
  parsed[["results"]] <- mutate_(parsed[["results"]],
                                 text_md5 = ~digest::digest(text, algo = "md5"))

  parsed[["properties"]] <- mutate_(parsed[["properties"]],
                                 text_md5 = ~digest::digest(text, algo = "md5"))

  # done!
  parsed

}


R/geoparser_query.R
#' Issue a geoparser query
#'
#' The function calls the \code{geoparser.io} API which identifies places
#' mentioned in the input text (in English), disambiguates those places, and
#' returns data about the places found in the text.
#'
#' @importFrom httr GET content
#' @importFrom jsonlite fromJSON
#'
#' @param text_input A vector of strings, where each size must be smaller than 8 KB. See
#'  \code{nchar(text_input, type = "bytes")}.
#' @param key Your \code{geoparser.io} key.
#'
#' @details To get an API key, you need to register at
#' \url{https://geoparser.io/pricing.html}.
#' With an hobbyist account, you can make up to 1,000 calls a month to the API.
#' For ease of use, save your API key as an environment variable as described
#' at \url{https://stat545-ubc.github.io/bit003_api-key-env-var.html}.
#'
#' The package will conveniently look for your API key using
#' \code{Sys.getenv("GEOPARSER_KEY")} so if your API key is an environment
#' variable called "\code{GEOPARSER_KEY}" you don't need to input it manually.
#'
#' \code{Geoparser.io} works best on complete sentences in English. If you
#' have a very short text, such as a partial address like
#' "\code{Auckland New Zealand}," you probably want to use a geocoder tool
#' instead of a geoparser. In R, you can use the
#' \href{https://cran.r-project.org/package=opencage}{opencage package} for
#' geocoding!
#'
#' @return A list of 2 \code{data.frame}s (\code{dplyr::tbl_df}s). The first
#' \code{data.frame} is called \code{properties} and contains:
#' \itemize{
#' \item \code{apiVersion} the version of the API currently in use
#' \item \code{source} the source of authority for the results
#' \item \code{id} the unique id of the query
#' \item \code{text_md5} MD5 hash of the text that was sent to the API.
#' }
#' The second \code{data.frame} contains the results and is called results.
#' If there are no results it is totally empty, otherwise:
#' \itemize{
#' \item \code{country} \href{https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2}{ISO-3166 2-letter country code}
#' for the country in which this place is located or \code{NULL} for features
#' outside any sovereign territory.
#' \item \code{confidence} A confidence score produced by the place
#' name disambiguation algorithm. \emph{Currently returns a placeholder value;
#' subject to change}.
#' \item \code{name} Best name for the specified location, with a
#' preference for official/short name forms (e.g., "\code{New York}" over
#' "\code{NYC}," and "\code{California}" over "\code{State of California}"),
#' which may be different from exactly what appears in the text.
#' \item \code{admin1} A code representing the state/province-level
#' administrative division containing this place. (From \code{GeoNames.org}:
#' \emph{"Most adm1 are FIPS codes. ISO codes are used for US, CH, BE and ME.
#' UK and Greece are using an additional level between country and FIPS code.
#' The code '\code{00}' stands for general features where no specific adm1 code
#' is defined.}")
#' \item \code{type} A text description of the geographic feature
#' type — \emph{see \url{http://www.geonames.org/} for a complete list. Subject to
#' change}.
#' \item \code{geometry.type} Type of the geographical feature, e.g.
#' "\code{Point}".
#' \item \code{longitude} Longitude.
#' \item \code{latitude} Latitude.
#' \item \code{reference1} Start (index of the first character in the place
#' reference) --  \emph{each reference to this place name found in the input
#' text is on one distinct line}.
#' \item \code{reference2} End (index of the first character after the place
#' reference) --  \emph{each reference to this place name found in the input
#' text is on one distinct line}.
#' \item \code{text_md5} MD5 hash of the text that was sent to the API.
#' }
#' @references \href{https://geoparser.io/docs.html}{\code{Geoparser.io}
#' documentation}
#' @export
#'
#' @examples
#' \dontrun{
#' geoparser_q(text_input = "I was born in Vannes but I live in Barcelona.")
#' geoparser_q(text_input = c("I was born in Vannes but I live in Barcelona.",
#' "France is the most beautiful place in the world."))
#' }
geoparser_q <- function(text_input,
                        key = geoparser_key()){

  # check arguments
  geoparser_query_check(text_input, key)

  result <- lapply(text_input, total, key = key)


  list(results = do.call("rbind", lapply(result, "[[", "results")),
       properties = do.call("rbind", lapply(result, "[[", "properties")))

}
